src/jvm/main/org/jetbrains/kotlinx/lincheck/trace/TraceReporter.kt

```diff
@@ -153,7 +153,10 @@ internal class TraceReporter(
                 (it is MethodReturnTracePoint) ||
                 it is SpinCycleStartTracePoint
             }
-            val isThreadJoinSwitch = (remainingTracePoints.firstOrNull()?.isThreadJoin() == true)
+            val isThreadJoinSwitch = (movedTracePoints.lastOrNull()?.isThreadJoin() == true) &&
+                // check that Thread.join() call has return value set, otherwise it is a hung join
+                (movedTracePoints.last() as MethodCallTracePoint).returnedValue != ReturnedValueResult.NoValue
```

Comment:
Is there a test that distinguishes these behaviours?

Reply:
`LivelockRunConcurrentRepresentationTest` distinguish this case. 
In fact, it is because this test was failing I had to add this line.